extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js

```diff
@@ -169,32 +176,99 @@ export class AmpGwdRuntimeService {
   }
 
   /**
-   * Stops animations on the previously-active page and starts them on the
-   * newly-active page.
-   * @param {number} index The index of the newly-active slide.
+   * Handles a page switch by resetting animations and goto counters on the
+   * currently-active page and starting animations on the new page.
+   * @param {number} index The index of the newly-active page (a slide in the
+   *     pagedeck amp-carousel).
    */
   setCurrentPage(index) {
-    // Turn off animations on the previously-active page, if there was one.
+    const gwdPages = this.ampdoc_.getRootNode().querySelectorAll(
+        `.${escapeCssSelectorIdent(GWD_PAGE_WRAPPER_CLASS)}`);
+
+    if (gwdPages.length == 0) {
+      // The document has no pages.
+      return;
+    }
+
+    // Deactivate the outgoing current page, if there is one.
     // TODO(sklobovskaya): Decide if it's worth just storing the index.
     const currentPageEl = this.ampdoc_.getRootNode().querySelector(
         `.${escapeCssSelectorIdent(GWD_PAGE_WRAPPER_CLASS)}.${
           escapeCssSelectorIdent(PlaybackCssClass.PLAY)
         }`);
 
     if (currentPageEl) {
-      currentPageEl.classList.remove(PlaybackCssClass.PLAY);
+      this.deactivatePage_(currentPageEl);
     }
 
     // Activate animations on the new current page.
-    const gwdPages = this.ampdoc_.getRootNode().querySelectorAll(
-        `.${escapeCssSelectorIdent(GWD_PAGE_WRAPPER_CLASS)}`);
     const newPageEl = gwdPages[index];
 
     if (newPageEl) {
-      newPageEl.classList.add(PlaybackCssClass.PLAY);
+      this.activatePage_(newPageEl);
+    } else {
+      user().error(LOG_ID, 'Could not find page with index ' + index + '.');
     }
   }
 
+  /**
+   * Sets a page as the current active page by enabling animations on it.
+   * Animations are prevented from running on inactive pages.
+   * @param {!Element} pageEl
+   * @private
+   */
+  activatePage_(pageEl) {
+    pageEl.classList.add(PlaybackCssClass.PLAY);
+  }
+
+  /**
+   * Sets a page inactive by disabling all animations and resetting all
+   * animation state (such as goto counters) on all elements within the page.
+   * @param {!Element} pageEl
+   * @private
+   */
+  deactivatePage_(pageEl) {
+    // Cancel and disable all animations on the page.
+    pageEl.classList.remove(PlaybackCssClass.PLAY);
+
+    // Reset other animation state on the page and all descendants.
+    const pageAndDescendants =
+        Array.prototype.slice.call(pageEl.querySelectorAll('*'), 0);
```

Comment:
1. Use `scopedQuerySelectorAll` from `src/dom.js`.
2. The performance advantage of using Array.call under a `NodeList` context vs. converting to an array is negligible. `toArray` from `src/types.js` should be fine.

Overall, I think this is more expressive:

```javascript
const pageAndDescendants = [pageEl].concat(scopedQuerySelectorAll(pageEl, '*'));

pageAndDescendants.forEach(el =>
  this.resetAnimatedElement_(el));

// Cancel and disable all animations on the page.
pageEl.classList.remove(PlaybackCssClass.PLAY);
```

Reply:
Thanks, I didn't know about toArray. Implemented your suggestion.

Reply:
Update: Linter says scopedQuerySelectorAll is not necessary here, so I kept querySelectorAll.